# --- Day 3: Mull It Over ---
## part 1
"Our computers are having issues, so I have no idea if we have any Chief Historians in stock! You're welcome to check the warehouse, though," says the mildly flustered shopkeeper at the North Pole Toboggan Rental Shop. The Historians head out to take a look.

The shopkeeper turns to you. "Any chance you can see why our computers are having issues again?"

The computer appears to be trying to run a program, but its memory (your puzzle input) is corrupted. All of the instructions have been jumbled up!

It seems like the goal of the program is just to multiply some numbers. It does that with instructions like mul(X,Y), where X and Y are each 1-3 digit numbers. For instance, mul(44,46) multiplies 44 by 46 to get a result of 2024. Similarly, mul(123,4) would multiply 123 by 4.

However, because the program's memory has been corrupted, there are also many invalid characters that should be ignored, even if they look like part of a mul instruction. Sequences like mul(4*, mul(6,9!, ?(12,34), or mul ( 2 , 4 ) do nothing.

For example, consider the following section of corrupted memory:

`xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))`

Only the four highlighted sections are real mul instructions. Adding up the result of each instruction produces 161 (2*4 + 5*5 + 11*8 + 8*5).

Scan the corrupted memory for uncorrupted mul instructions. What do you get if you add up all of the results of the multiplications?

In [ ]:
from aocd import get_data
import random
input_data = get_data(day=3, year=2024)
random.sample(input_data.splitlines(),1)

[">$~when()^;mul(914,586)why()$mul(383,297)when()when()select(85,106)select()!$;mul(566,651) mul(928,762)mul(302,399))( !)<mul(926,250))(mul(842,947)}from()mul(53,294)&mul(784,942)mul(614,108))[from()&,;mul(67,626)%-[mul(516,770)/<?}#$;({from()mul(77,888)how()what(){@/mul(392,76)how()^when()[<'where(){mul(357,569)$mul(866,257)^?what())*-who(87,183)<mul(371,728){'what()!+mul(359,893)!;who()^{)%where()/%mul(350,660)/from()select()%^+)@when()mul(331,754)} * )(mul(98,91)[/*from()?-do()^:,when())when()$~who()mul(457,779)mul(825,607)(from()from()<mul(888,364)#how(){(<})what()[mul(356,842);>-{[%;:$}mul(973,195)* #$from()who()/mul(383,305):]~}+'^what(92,64)mul(360,235who();mul(186,635)[}?,/why()#what()&select()mul(249,256)&/ mul(851,885)?:mul(797@,mul(87,662)what(309,502)$#}-mul(144,693),</from(621,475);who()+-;mul(489,471)select()where(){mul(101,882)mul(399,691) -:mul(203-how()from()when()(}'!from()>mul(148,850)*mul(781,765)select()do()?> @*mul(262^who()[&!mul(557,94)select()don't()'from()mul

In [ ]:
line = 'xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))'
line

'xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))'

In [ ]:
import re

**AI Prompt**

I want to use regex to extract mul(d,d) out of a line. Can you teach my the syntax? Im not too familiar with regex. I know I can use \d for digit.

**AI Response**

I'll help you learn regex step by step for this specific problem. Let's break it down:

1. First, we need to match "mul" literally
2. Then we need an opening parenthesis
3. Then 1-3 digits
4. Then a comma
5. Then 1-3 digits again
6. Finally a closing parenthesis

Would you like me to explain how to write each part of this pattern? We can test them one at a time to make sure you understand each piece.

Let me know if you'd like to start with the first couple of parts and we'll build from there.

**AI Prompt**

Which re.method will I be using and what are the inputs?

**AI Response**

For this task, `re.findall()` would be most useful. It returns all non-overlapping matches of the pattern in the string as a list.

The basic syntax is:
```python
re.findall(pattern, string)
```

Where:
- `pattern` is your regex pattern (we'll build this together)
- `string` is the text you want to search (in your case, each line of input)

Would you like to try a simple test with just the first part of our pattern to see how it works?

In [ ]:
re.findall('mul\(\d,\d\)', line)

['mul(2,4)', 'mul(5,5)', 'mul(8,5)']

**AI Prompt**

Like I said, Im really new to this. I was trying to extract for example one digit examples. I remember somethibg about escaping special chars like brackets.

**AI Response**

Ah, let's start with the absolute basics! You're right about escaping special characters. In regex:

1. To match a literal parenthesis `(` or `)`, we need to use `\(` or `\)`
2. To match a single digit, we use `\d`

Let's try matching just "mul" followed by one digit. Would you like to try writing this pattern yourself? Or would you like me to show you how?

Remember, we can test our pattern on the example line:
```
xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))
```

In [ ]:
re.findall('mul\(\d', line)

['mul(2', 'mul(5', 'mul(3', 'mul(1', 'mul(8']

**AI Prompt**

Like that?

**AI Response**

Yes, exactly! You've got it! Your pattern 'mul\(\d' successfully matches:
- 'mul' literally
- An opening parenthesis (escaped with \)
- A single digit (\d)

Would you like to try extending this pattern to also match the comma after the digit? Remember, a comma is just a literal character, so you don't need to escape it.

In [ ]:
re.findall('mul\(\d,', line)

['mul(2,', 'mul(5,', 'mul(8,']

**AI Prompt**

ok done

**AI Response**

Great! Now can you try adding one more digit after the comma? You'll use the same \d pattern we used for the first digit. Then we'll just need the closing parenthesis to complete the pattern.

Would you like to try that?

In [ ]:
re.findall('mul\(\d,\d\)', line)

['mul(2,4)', 'mul(5,5)', 'mul(8,5)']

**AI Prompt**

yep that's what I had orginally

**AI Response**

You're right! Now, to handle numbers that might be 1-3 digits long (like 123), we need to modify the \d part. In regex, we can specify "1 to 3 repetitions" using {1,3} after a pattern.

Would you like to try modifying your pattern to handle multi-digit numbers?

In [ ]:
re.findall('mul\(\d{1,3},\d{1,3}\)', line)

['mul(2,4)', 'mul(5,5)', 'mul(11,8)', 'mul(8,5)']

**AI Prompt**

like that? Although our test case is not too great b/c it does not have 3 digit examples or 4 digits etc.

**AI Response**

Exactly right! And that's a great observation about testing. Would you like to create a test string with some more complex cases to verify our pattern? We could include:
- 3-digit numbers (like 123)
- 4-digit numbers (which shouldn't match)
- Some invalid formats

This way we can be sure our pattern is working correctly. Would you like to try creating such a test case?

**AI Prompt**

You can create one with some edge cases etc

**AI Response**

Let's create a test string with various cases:

```python
line = ">$~when()^;mul(914,586)whymul(123,456)mul(350,660)/from()select()%^+)@when()mumul(1,2)ul(994,907)$mul(510,490)&where()mul(758mul(1234,56)479)~~^^-mul(88mul(12,5678) ^why()&when()$select()~%mul(mul(12.3,45)o()$what()':how(573,320)mul(85,2mul(12,)[@when(32,211)@mul(544,929)/select()from()mul(,34)l(787,474)&+where()mumul12,34)mul(12:34)"

In [ ]:
line = ">$~when()^;mul(914,586)whymul(123,456)mul(350,660)/from()select()%^+)@when()mumul(1,2)ul(994,907)$mul(510,490)&where()mul(758mul(1234,56)479)~~^^-mul(88mul(12,5678) ^why()&when()$select()~%mul(mul(12.3,45)o()$what()':how(573,320)mul(85,2mul(12,)[@when(32,211)@mul(544,929)/select()from()mul(,34)l(787,474)&+where()mumul12,34)mul(12:34)uwerhf90tmul(99, 2)mul(12,13)"
line

">$~when()^;mul(914,586)whymul(123,456)mul(350,660)/from()select()%^+)@when()mumul(1,2)ul(994,907)$mul(510,490)&where()mul(758mul(1234,56)479)~~^^-mul(88mul(12,5678) ^why()&when()$select()~%mul(mul(12.3,45)o()$what()':how(573,320)mul(85,2mul(12,)[@when(32,211)@mul(544,929)/select()from()mul(,34)l(787,474)&+where()mumul12,34)mul(12:34)uwerhf90tmul(99, 2)mul(12,13)"

In [ ]:
re.findall('mul\(\d{1,3},\d{1,3}\)', line)

['mul(914,586)',
 'mul(123,456)',
 'mul(350,660)',
 'mul(1,2)',
 'mul(510,490)',
 'mul(544,929)',
 'mul(12,13)']

In [ ]:
assert re.findall('mul\(\d{1,3},\d{1,3}\)', line) == ['mul(914,586)', 'mul(123,456)','mul(350,660)', 'mul(1,2)','mul(510,490)','mul(544,929)', 'mul(12,13)']

**AI Prompt**

Okay looking good right?

**AI Response**

Yes, looking very good! Our pattern successfully:
- Matches valid mul() with 1-3 digits on each side
- Ignores cases with 4+ digits
- Ignores broken syntax (like missing numbers, decimal points, or spaces)
- Ignores cases where 'mul' is part of another word

Are you ready to apply this to your original puzzle input to solve the problem?

**AI Prompt**

Is there a way to use regex to just extract the numbers into a list of tuples?
I know I could strip the `mul(` and trailing `)` manually but there must be a more re way right?

**AI Response**

Yes! In regex, you can use parentheses () to create "capture groups" - anything that matches inside these groups will be extracted. We put these around the parts we want to capture.

For example, if we modify our pattern to:
```python
'mul\((\d{1,3}),(\d{1,3})\)'
```

The parentheses around `\d{1,3}` will capture just the numbers. Would you like to try this and see what it returns?

In [ ]:
re.findall('mul\((\d{1,3}),(\d{1,3})\)', line)

[('914', '586'),
 ('123', '456'),
 ('350', '660'),
 ('1', '2'),
 ('510', '490'),
 ('544', '929'),
 ('12', '13')]

In [ ]:
[(int(a), int(b)) for a,b in re.findall('mul\((\d{1,3}),(\d{1,3})\)', line)]

[(914, 586), (123, 456), (350, 660), (1, 2), (510, 490), (544, 929), (12, 13)]

In [ ]:
[int(a) * int(b) for a,b in re.findall('mul\((\d{1,3}),(\d{1,3})\)', line)]

[535604, 56088, 231000, 2, 249900, 505376, 156]

In [ ]:
sum([int(a) * int(b) for a,b in re.findall('mul\((\d{1,3}),(\d{1,3})\)', line)])

1578126

Let's test with original sample test data.

In [ ]:
line = 'xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5)'
sum([int(a) * int(b) for a,b in re.findall('mul\((\d{1,3}),(\d{1,3})\)', line)])

161

In [ ]:
answer = 0
for line in input_data.splitlines():
    line_sum = sum([int(a) * int(b) for a,b in re.findall('mul\((\d{1,3}),(\d{1,3})\)', line)])
    answer += line_sum
answer

173731097

**AI Prompt**

Looks good right? Ready for submission?

**AI Response**

Yes, it looks perfect! Your solution:
1. Correctly extracts number pairs using capture groups
2. Converts strings to integers
3. Multiplies each pair
4. Sums up all the results
5. Works correctly on the example data (getting 161)

And you've processed all lines in the input data. The answer 173731097 should be ready for submission!

Would you like to try part 2 of the puzzle next?

## --- Part Two ---

As you scan through the corrupted memory, you notice that some of the conditional statements are also still intact. If you handle some of the uncorrupted conditional statements in the program, you might be able to get an even more accurate result.

There are two new instructions you'll need to handle:

The do() instruction enables future mul instructions.
The don't() instruction disables future mul instructions.
Only the most recent do() or don't() instruction applies. At the beginning of the program, mul instructions are enabled.

For example:

`xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))`

This corrupted memory is similar to the example from before, but this time the mul(5,5) and mul(11,8) instructions are disabled because there is a don't() instruction before them. The other mul instructions function normally, including the one at the end that gets re-enabled by a do() instruction.

This time, the sum of the results is 48 (2*4 + 8*5).

Handle the new instructions; what do you get if you add up all of the results of just the enabled multiplications?

**AI Prompt**

So the first thing is for me to understand what part 2 is asking.

- all the `mul` after a `don't()` are disabled. 
- They are only enabled if there is a `do()` before it.
- only the most recent do/dont before a mul matters.

I got that right, correct? If so, I think I need you to make a good test example like we did before.

**AI Response**

Yes, you've understood it perfectly! Let's create a test string with various cases to verify our logic. 

```python
line = "mul(2,3)don't()mul(4,5)mul(6,7)do()mul(8,9)don't()mul(10,11)do()mul(12,13)mul(14,15 don't()mul(16,17)"
```

**AI Prompt**

I'm thinking I want to prepend a `do()` to the beginning of the line and add a `don't()` at the end.

`line = 'do()' + line + "don't()"`

Then what I want to do is take out all the symbols in the sequence between each do() and don't(). 
I will do this for each such sequence.
Then I will be left with Sequences I can concat. Then I can use solution from part 1.
Does that make sense? Im sure there are other ways but thats the first thing that came to mind.

**AI Response**

That's a clever approach! Let me break down what I understand from your plan:

1. Add `do()` at start and `don't()` at end to standardize the format
2. Find all sequences between `do()` and `don't()`
3. Clean up each sequence (remove non-mul stuff)
4. Concatenate all valid sequences
5. Use part 1's solution on the result

This should work! Would you like to try implementing it step by step? We could start by writing a regex to find the sequences between `do()` and `don't()`. Or would you prefer to try a different approach?

In [ ]:
line = "mul(2,3)don't()mul(4,5)mul(6,7)do()mul(8,9)don't()mul(10,11)do()mul(12,13)mul(14,15 don't()mul(16,17)"
line = ' do() ' + line + " don't() "
line

" do() mul(2,3)don't()mul(4,5)mul(6,7)do()mul(8,9)don't()mul(10,11)do()mul(12,13)mul(14,15 don't()mul(16,17) don't() "

**AI Prompt**

What is the regex which capture everything? Even new lines. In general Im asking.

**AI Response**

The dot (.) in regex normally matches any character except newlines. To match everything including newlines, you can use:

- `.*?` for non-greedy matching
- `.*` for greedy matching

But to make the dot match newlines too, you need to add the flag `re.DOTALL` or use the pattern `[\s\S]*`

Would you like me to explain the difference between these approaches, or shall we try using one of them in your solution?

In [ ]:
re.findall("do\(\)[\s\S]*don\'t\(\)", line)

["do() mul(2,3)don't()mul(4,5)mul(6,7)do()mul(8,9)don't()mul(10,11)do()mul(12,13)mul(14,15 don't()mul(16,17) don't()"]

**AI Prompt**

How come its only finding one thing?

**AI Response**

Ah, that's because we need to make the match non-greedy! Right now it's being greedy, which means it's matching from the first `do()` all the way to the last `don't()` it can find.

Try adding a `?` after the `*` to make it non-greedy:
```python
re.findall("do\(\)[\s\S]*?don\'t\(\)", line)
```

This will match the shortest possible string between each `do()` and `don't()` instead of the longest possible string. Want to try it?

In [ ]:
re.findall("do\(\)[\s\S]*?don\'t\(\)", line)

["do() mul(2,3)don't()",
 "do()mul(8,9)don't()",
 "do()mul(12,13)mul(14,15 don't()"]

In [ ]:
re.findall("do\(\)([\s\S]*?)don\'t\(\)", line)

[' mul(2,3)', 'mul(8,9)', 'mul(12,13)mul(14,15 ']

In [ ]:
' '.join(re.findall("do\(\)([\s\S]*?)don\'t\(\)", line))

' mul(2,3) mul(8,9) mul(12,13)mul(14,15 '

In [ ]:
line = "xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))"
print(line)
def part2_answer(line):
    line = ' do() ' + line + " don't() "
    line = ' '.join(re.findall("do\(\)([\s\S]*?)don\'t\(\)", line))
    return sum([int(a) * int(b) for a,b in re.findall('mul\((\d{1,3}),(\d{1,3})\)', line)])
part2_answer(line)

xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))


48

In [ ]:
part2_answer(input_data)

93729253